### PR TITLE
feat: timelock multi-token reward bond

### PR DESCRIPTION
**Pull request by CjHare**
### Purpose for this PR

Closes https://github.com/windranger-io/windranger-treasury/issues/254

### CODE REVIEW

File: contracts/bond/TimeLockMultiRewardBond.sol [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+import "@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol";
+import "@openzeppelin/contracts-upgradeable/token/ERC20/IERC20Upgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/security/PausableUpgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/utils/ContextUpgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/utils/structs/EnumerableSetUpgradeable.sol";
+
+/**
+ * @title Multiple reward with time lock support.
+ *
+ * @notice Supports multiple ERC20 rewards with optional an optional time lock on pull based claiming.
```
Comment (aodhgan) [inline contracts/bond/TimeLockMultiRewardBond.sol:12]: nit: typo

File: contracts/bond/TimeLockMultiRewardBond.sol [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+        emit UpdateRewardTimeLock(tokens, timeLock);
+    }
+
+    /**
+     * @notice The time at which the debt tokens are redeemable.
+     *
+     * @dev Until a redemption time is set, no rewards are claimable.
+     */
+    function _setRedemptionTimestamp(uint256 timestamp) internal whenNotPaused {
+        require(timestamp != 0, "Rewards: zero is invalid time");
```
Comment (aodhgan) [inline contracts/bond/TimeLockMultiRewardBond.sol:187]: do we want to require if this is in the future? `timestamp > block.timestamp`

File: contracts/bond/TimeLockMultiRewardBond.sol [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+                _rewardPool[tokens].amount,
+            "Rewards: not enough held"
+        );
+    }
+
+    /**
+     * @notice Overwrites the existing time lock for rewards from a single ERC20.
+     *
+     * @param tokens ERC20 rewards already registered.
+     * @param timeLock ms to lock rewards after redemption is allowed.
```
Comment (aodhgan) [inline contracts/bond/TimeLockMultiRewardBond.sol:168]: nit: seconds vs miliiseconds? had to google this one lol

File: contracts/bond/TimeLockMultiRewardBond.sol (hunk 1 above)
Comment (CjHare) [reply contracts/bond/TimeLockMultiRewardBond.sol:12]: good spot 👍

File: contracts/bond/TimeLockMultiRewardBond.sol (hunk 3 above)
Comment (CjHare) [reply contracts/bond/TimeLockMultiRewardBond.sol:168]: Are you saying the Linux timestamp is in seconds not milliseconds?

File: contracts/bond/TimeLockMultiRewardBond.sol (hunk 2 above)
Comment (CjHare) [reply contracts/bond/TimeLockMultiRewardBond.sol:187]: That would be a more sensible comparison ...wonder if Slither is going to complain about it 🤔

Comment (CjHare) [reply contracts/bond/TimeLockMultiRewardBond.sol]: https://docs.soliditylang.org/en/latest/units-and-global-variables.html?highlight=block.timestamp#block-and-transaction-properties

Solidity time is indeed in seconds 😶‍🌫️

**Pull request merged**